astropy/cosmology/funcs.py

```diff
@@ -255,3 +276,105 @@ def f(z):
+as a bracket and then truncating the object array to remove the placeholder.
+This can be seen in the following example:
+
+>>> bracket=np.array([(1.0, 1.2),(2.0, 2.5), ()], dtype=object)[:-1]
+>>> z_at_value(Planck18.angular_diameter_distance, 1500 * u.Mpc,
+...            bracket=bracket)  # doctest: +SKIP
+array([0.68044452, 3.7823268])
+"""
+
+
+@format_doc(_docstring_z_at_value + _docstring_z_at_value_vector_examples,
+            func_name="z_at_value", name_space="",
+            fval_type="scalar or vector",
+            z_type="float or array-like['dimensionless']",
+            maxfun_type="int or array-like",
+            bracket_type="sequence or object array[sequence]",
+            return_type="scalar or array", raises_docs=_raises_docs)
+def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500,
+               method='Brent', bracket=None, verbose=False):
+    """``z_at_scalar_value`` for vectorized inputs."""
+    # `fval` can be a Quantity, which isn't (yet) compatible w/ `numpy.nditer`
+    # so we strip it of units for broadcasting and restore the units when
+    # passing the elements to `z_at_scalar_value`.
+    fval = np.asanyarray(fval)
+    unit = getattr(fval, 'unit', 1)  # can be unitless
+
+    # bracket must be an object array (assumed to be correct) or a 'scalar'
+    # bracket: 2 or 3 elt sequence
+    if not isinstance(bracket, np.ndarray):  # 'scalar' bracket
+        if bracket is not None and len(bracket) not in (2, 3):
+            raise ValueError("`bracket` is not an array "
+                             "nor a 2 (or 3) element sequence.")
+        else:  # munge bracket into a 1-elt object array
+            bracket = np.array([bracket, ()], dtype=object)[:1].squeeze()
+    if bracket.dtype != np.object_:
+        raise TypeError(f"`bracket` has dtype {bracket.dtype}, not 'O'")
+
+    # make multi-dimensional iterator for all but `method`, `verbose`
+    # TODO: figure out 'reduce_ok', so scalar returns scalar
+    with np.nditer(
+        [fval, zmin, zmax, ztol, maxfun, bracket, None],
+        flags = ['refs_ok'],
+        op_flags = [*[['readonly']] * 6,  # ← inputs  output ↓
+                    ['writeonly', 'allocate', 'no_subtype']],
+        op_dtypes = (*(None,)*6, fval.dtype),
+        casting="no",
+    ) as it:
+        for fv, zmn, zmx, zt, mfe, bkt, zs in it:  # ← eltwise unpack & eval ↓
+            zs[...] = z_at_scalar_value(func, fv * unit, zmin=zmn, zmax=zmx,
+                                        ztol=zt, maxfun=mfe, bracket=bkt.item(),
+                                        # not broadcasted
+                                        method=method, verbose=verbose)
+        # since bracket is an object array, the output will be too, so it is
+        # cast to the same type as the function value.
+        result = it.operands[-1]  # zs
+
+    # degrade to scalar if 0d array  # TODO: rm when 'reduce_ok' works
+    return result if result.shape else result.item()
```

Comment:
I'd tend to do the scalar case with `[()]` - this produces a numpy scalar rather than a python scalar, and will likely behave better for things like Masked numbers.

Reply:
As in ``return result[()] if result.shape else result.item()`` ?

Reply:
Yup. That works. Thanks!

Reply:
No, the other way around: `return result if result.shape else result[()]`